proposals/2021/08-25_repo-structure.md

```diff
@@ -0,0 +1,320 @@
+
+## Problem 1
+
+The first problem with this structure is ensuring consistency, both currently and going forward, among the repos in the [go-vela](https://github.com/go-vela) org.
+
+Currently, not all "core" repos follow the `pkg-<capability>` naming convention. i.e. [compiler](https://github.com/go-vela/compiler)
+
+Also, some "core" repos have the packages for different capabilities in the repo itself. i.e. [server](https://github.com/go-vela/server)
+
+If we plan to keep the `pkg-<capability>` structure, those should be pulled out into unique repos following that structure:
+
+* [server/database](https://github.com/go-vela/server/tree/master/database) -> `pkg-database`
+* [server/secret](https://github.com/go-vela/server/tree/master/secret) -> `pkg-secret`
+* [server/source](https://github.com/go-vela/server/tree/master/source) -> `pkg-source`
+
+## Problem 2
+
+The second problem with this structure is the increased burden, especially for new users, trying to contribute functionality.
+
+Some, or most, of the "core" repos have dependencies on one or more repos.
+
+This means that in order to make certain changes to the product, you have to introduce code changes to multiple repos.
+
+To provide an example with the current structure, we'll reference a recent feature that allows you to customize the pipeline type for a repo.
+
+This functionality was added as a part of [go-vela/community#326](https://github.com/go-vela/community/issues/326).
+
+To achieve this, a new `pipeline_type` field was added to the database which required code changes for multiple repos:
+
+* [types](https://github.com/go-vela/types/pull/188)
+* [compiler](https://github.com/go-vela/compiler/pull/199)
+* [server](https://github.com/go-vela/server/pull/444)
+* [ui](https://github.com/go-vela/ui/pull/421)
+
+This also increases the complexity when attempting to functionally test code changes for new logic.
+
+This is due to needing to modify each repo locally with the code changes for the functionality.
+
+Once that is complete, the application must be rebuilt from the local code changes for each repo.
+
+## Problem 3
+
+The third problem with this structure is level of overhead when attempting to publish new releases for the product.
+
+The way the repos are setup today, there is a dependency tree that must be followed in order to properly release the product.
+
+The way the repos must be released is in groups that use the following order:
+
+1. [types](https://github.com/go-vela/types), [ui](https://github.com/go-vela/ui)
+2. [compiler](https://github.com/go-vela/compiler), [mock](https://github.com/go-vela/mock)
+3. [pkg-runtime](https://github.com/go-vela/pkg-runtime), [sdk-go](https://github.com/go-vela/sdk-go)
+4. [pkg-executor](https://github.com/go-vela/pkg-executor), [pkg-queue](https://github.com/go-vela/pkg-queue)
+5. [cli](https://github.com/go-vela/cli), [server](https://github.com/go-vela/server), [worker](https://github.com/go-vela/worker)
```

Comment:
So, if I'm understanding this correctly with this proposal we'd end up eliminating steps 2 & 4 and step 3 would only be the `sdk-go`? Seems like quite the improvement 🎉

Reply:
That is correct 😃